### ISSUE TITLE
[Object Store Access] Context Deadline exceeded after upgrading to v0.32.4 from v0.31.0

### ISSUE THREAD

**Issue by jnyi** [thanos-io/thanos#6785]
**Thanos, Prometheus and Golang version used**: 
* v0.31.0 --> v0.32.4

**Object Storage Provider**: AWS s3

**What happened**: We used k8s and assume role for accessing AWS s3, it worked fine with thanos v0.31.0, after upgrading to v0.32.4, we start seeing compactor complaining, seems regression related to permission.

**What you expected to happen**: thanos compactor can talk to AWS s3 using assumeRole.

**How to reproduce it (as minimally and precisely as possible)**: normal s3 configuration without users/password, assume role of k8s worker to s3

**Full logs to relevant components**:

 
```
ts=2023-10-09T18:20:13.796433303Z caller=blocks_cleaner.go:44 level=info name=thanos-compactor msg="started cleaning of blocks marked for deletion"
ts=2023-10-09T18:20:13.797847735Z caller=blocks_cleaner.go:58 level=info name=thanos-compactor msg="cleaning of blocks marked for deletion done"
ts=2023-10-09T18:24:50.467766592Z caller=runutil.go:100 level=error name=thanos-compactor msg="function failed. Retrying in next tick" err="BaseFetcher: iter bucket: Get \"https://<redacted bucket>.s3.dualstack.us-west-2.amazonaws.com/?continuation-token=1SB%2BPSCTBxd2pyekgPnjuiCyJtzT1KzRQ8qEs8tNaBdv8HwvEXj7IgJrXZdEGs6m4gYJvI%2B1yKU7Pr57oxgIsufA2RdL0wtRHOhFqyckpiPpFq7JKDCa2npe1SGnPx7vOdU917CidMIt6MWcjlFD2GQ%3D%3D&delimiter=&encoding-type=url&fetch-owner=true&list-type=2&prefix=thanos%2Foregon-dev%2F\": context deadline exceeded"
ts=2023-10-09T18:24:50.46789616Z caller=compact.go:597 level=error name=thanos-compactor msg="retriable error" err="BaseFetcher: iter bucket: Get \"https://<redacted bucket>.s3.dualstack.us-west-2.amazonaws.com/?continuation-token=1SB%2BPSCTBxd2pyekgPnjuiCyJtzT1KzRQ8qEs8tNaBdv8HwvEXj7IgJrXZdEGs6m4gYJvI%2B1yKU7Pr57oxgIsufA2RdL0wtRHOhFqyckpiPpFq7JKDCa2npe1SGnPx7vOdU917CidMIt6MWcjlFD2GQ%3D%3D&delimiter=&encoding-type=url&fetch-owner=true&list-type=2&prefix=thanos%2Foregon-dev%2F\": context deadline exceeded"
```

**Anything else we need to know**:

**Comment by yeya24** [comment]
@jnyi Is this reproducible every time? It looks like a small network issue that should happen rarely.

**Comment by jnyi** [comment]
it is reproducible, we are using IAM assume role in AWS instead of direct username/password, roll back to older version fixed this issue.

**Comment by Pacobart** [comment]
Also notice in the config that the user set the s3 endpoint as `s3.us-west-2.amazonaws.com` but the logs show `s3.dualstack.us-west-2.amazonaws.com`. This another issue https://github.com/thanos-io/thanos/issues/6804

**Comment by yeya24** [comment]
Umm, @Pacobart did you see Thanos using dualstack endpoint only recently after the upgrade?
Looks like dualstack was used since 2019 https://github.com/minio/minio-go/pull/1055

**Comment by Pacobart** [comment]
> Umm, @Pacobart did you see Thanos using dualstack endpoint only recently after the upgrade? Looks like dualstack was used since 2019 [minio/minio-go#1055](https://github.com/minio/minio-go/pull/1055)

Great find! This is unfortunate they did that but at least I know now where it's coming from.

**Comment by fpetkovski** [comment]
Is there a way to skip old version when iterating objects in S3? Maybe that is something we can do on our side.

**Comment by jnyi** [comment]
Speaking about the object storage layout, taking Grafana Mimir as an example, they have layout breakdown by tenant while thanos today put all tenants + raw resolution + 5m/1h downsampled buckets in 1 places, will prefix them separately and only iterate related logic subpaths make it simpler?

**Comment by yeya24** [comment]
Like Cortex there is bucket index to help speed up the time when loading blocks. I think Thanos might be able to have similar things to avoid iterating the whole bucket every time.

**Comment by fpetkovski** [comment]
Bucket index would be a great addition, I wonder how hard it is to merge that code from Cortex. However, we do need a short term solution to unblock people from upgrading.

**Comment by MichaHoffmann** [comment]
Can we update a directory besides the blocks where just files with names of blocks are? everytime we add blocks we write a new file, everytime we delete stuff we delete the fiels there. then we only need to look there for blocks

**Comment by MichaHoffmann** [comment]
> Bucket index would be a great addition, I wonder how hard it is to merge that code from Cortex. However, we do need a short term solution to unblock people from upgrading.

for short term we can introduce a hidden flag to opt into the new behavior maybe and long term we can go towards a block index solution; wdyt?

**Comment by yeya24** [comment]
https://cortexmetrics.io/docs/blocks-storage/bucket-index/ Sharing Cortex bucket index doc.
Thanos might not need all the features from it but idea is the same

**Comment by michaelswierszcz** [comment]
--block-viewer.global.sync-block-timeout=5m
--block-viewer.global.sync-block-interval=1m

https://thanos.io/tip/components/compact.md/

changing these defaults on thanos compactor fixes the context deadline timeouts in my cluster

**Comment by BouchaaraAdil** [comment]
the same here i upgraded Thanos from `0.31.0 ` to `0.34.1`
and I do see a lot of :
```
level=error msg="retriable error" err="syncing metas: BaseFetcher: iter bucket: Get \"https://bucket_name.s3.dualstack.eu-west-1.amazonaws.com/?continuation-token=1CsmotG2%xxxxxxxxxxxxxxxxxx%xxxxxxxxx%xxxxxxxxx%3D&delimiter=&encoding-type=url&fetch-owner=true&list-type=2&prefix=\": context deadline exceeded"
```

**Comment by BouchaaraAdil** [comment]
I even tested `0.32.0` and it didn't work, faced the same issue ☝️ 
i reverted the compactor to 0.31.0 and kept other components at `0.34.1`

**Comment by BouchaaraAdil** [comment]
> --block-viewer.global.sync-block-timeout=5m
> --block-viewer.global.sync-block-interval=1m

i confirm these worked. thanks @michaelswierszcz
